storage/index.go

```diff
@@ -21,61 +21,260 @@
 package storage
 
 import (
+	"bytes"
 	"errors"
+	"sync"
+	"time"
 
 	"github.com/m3db/m3db/storage/index"
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment/mem"
 	"github.com/m3db/m3x/context"
 	"github.com/m3db/m3x/ident"
+
+	"github.com/uber-go/tally"
 )
 
 var (
-	errIndexingNotImplemented = errors.New("indexing is not implemented")
+	maxTime = time.Unix(1<<63-1, 0)
+
+	errDbIndexAlreadyClosed                      = errors.New("database index has already been closed")
+	errDbIndexUnableToWriteClosed                = errors.New("unable to write to database index, already closed")
+	errDbIndexUnableToQueryClosed                = errors.New("unable to query database index, already closed")
+	errDbIndexUnableToIndexWithReservedFieldName = errors.New("unable to index document due to usage of reserved fieldname")
+)
+
+type dbIndexState byte
+
+const (
+	dbIndexStateOpen dbIndexState = iota
+	dbIndexStateClosed
 )
 
+type dbIndexInsertMode byte
+
+// nolint
+const (
+	dbIndexInsertSync dbIndexInsertMode = iota
+	dbIndexInsertAsync
+)
+
+type dbIndexBlock struct {
+	segment    mem.Segment
+	expiryTime time.Time
+}
+
 type dbIndex struct {
-	opts Options
+	sync.RWMutex
+	insertMode dbIndexInsertMode
+	state      dbIndexState
+	active     dbIndexBlock
+
+	insertQueue databaseIndexInsertQueue
+	metrics     dbIndexMetrics
+	opts        index.Options
 }
 
-func newDatabaseIndex(o Options) (databaseIndex, error) {
-	return &dbIndex{
-		opts: o,
-	}, nil
+// nolint: deadcode
+func newDatabaseIndex(
+	fn newDatabaseIndexInsertQueueFn,
+	opts index.Options,
+) (databaseIndex, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	seg, err := mem.New(1, opts.MemSegmentOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	idx := &dbIndex{
+		insertMode: dbIndexInsertAsync,
+		active: dbIndexBlock{
+			segment:    seg,
+			expiryTime: maxTime, // FOLLOWUP(prateek): undo hard-coding to infinite retention
```

Comment:
Now that we are going to make this per namespace, can we just use the namespace retention options here?

Reply:
sure, will take as a ctor parameter and pass down.

Reply:
done